### DISCUSSION TITLE
Barrel files

### DISCUSSION THREAD

**Discussion by Jayllyz** [panva/jose#708] [Ideas]
Jose uses barel files that slow down performance, and causes large module graphs with unused modules.

For more information:
https://biomejs.dev/linter/rules/no-barrel-file/
https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7/

Suggestion:

Refactor codes to remove barel files, + setup a linter such as Biome or Eslint could help keep a clean code base.

**Comment by panva** [comment]
> 📖 tl;dr: Many projects are **littered** with files that **just** re-export other files.

By this definition, `jose` has exactly **one** "barrel" file and that's the index file. How would you suggest to structure the project instead?

**Reply by Jayllyz** [reply]
You can find a guide here, it explain how you can avoid barel files as lib author : https://thepassle.netlify.app/blog/practical-barrel-file-guide-for-library-authors

**Reply by panva** [reply]
I did introduce the pattern of individual exports with 3.x, it was not received well and developer tooling wasn't ready to use it so I went back to a simple well-established `import *` or `require()` in v4.x. Maybe the tooling landscape has improved since but I won't be going back to individual exports as the default now.

If you care deeply about the amount of code you import then you likely already have a build process which resolves the imports and tree-shakes what you don't use.

**Reply by wparad** [reply]
As a user, omg, it is so much better than 3.x, that was a nightmare. I don't know why anyone creates independent exports, that's just pain and suffering for everyone, please never go back to that.

**Reply by panva** [reply]
That being said, I don't think anything prohibits the module from having both a general `import * as jose from 'jose'` pattern as well as specific subpath module exports in addition

**Reply by panva** [reply]
https://github.com/panva/jose/releases/tag/v5.8.0
